File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.py
import sys
from collections import deque

T = int(input())

def sol(sx, sy, conv):
    q = deque()
    q.append([sx, sy])
    visited = [False] * (n+1)
    while q:
        x, y = q.popleft()

        if x == end_x and y == end_y and (abs(end_x - x) + abs(end_y - y)) <= 1000:
            print("happy")
            return

        for i in range(n+1):
            diff = abs(x-conv[i][0]) + abs(y-conv[i][1])
            if diff <= 1000 and not visited[i]:
                q.append([conv[i][0], conv[i][1]])
                visited[i] = True

    print("sad")
    return

for _ in range(T):
    n = int(input())
    start_x, start_y = map(int, input().split()) # 시작 위치
    conv = []
    for _ in range(n):
        x, y = map(int, input().split())
        conv.append([x, y])

    end_x, end_y = map(int, input().split())
    conv.append([end_x, end_y])
    sol(start_x, start_y, conv)

File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.java
import java.io.*;
import java.util.*;

public class Main {
    static int n;
    static Point[] stores;
    static Point end;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());

        while (T --> 0) {
            n = Integer.parseInt(br.readLine());
            StringTokenizer st = new StringTokenizer(br.readLine());
            int startX = Integer.parseInt(st.nextToken());
            int startY = Integer.parseInt(st.nextToken());

            stores = new Point[n];
            for (int i = 0; i < n; i++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                stores[i] = new Point(x, y);
            }

            st = new StringTokenizer(br.readLine());
            int endX = Integer.parseInt(st.nextToken());
            int endY = Integer.parseInt(st.nextToken());
            end = new Point(endX, endY);

            System.out.println(bfs(startX, startY));
        }

    }

    static String bfs(int sx, int sy) {
        Queue<Point> q = new LinkedList<>();
        boolean[] visited = new boolean[n];
        q.offer(new Point(sx, sy));

        while (!q.isEmpty()) {
            Point cur = q.poll();

            if (distance(cur, end) <= 1000) {
                return "happy";
            }

            for (int i = 0; i < n; i++) {
                if (!visited[i] && distance(cur, stores[i]) <= 1000) {
                    visited[i] = true;
                    q.offer(stores[i]);
                }
            }
        }

        return "sad";
    }

    static int distance(Point a, Point b) {
        return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
    }

    static class Point {
        int x, y;
        Point(int x, int y) { this.x = x; this.y = y;}
    }

}